src/rust/engine/rule_graph/README.md

```diff
@@ -0,0 +1,115 @@
+# Rule Graph Construction
+
+## Overview
+
+Build logic in [Pants](https://www.pantsbuild.org/) is declared using collections of `@rules` with recursively memoized and invalidated results. This framework (known as Pants' "Engine") has similar goals to Bazel's [Skyframe](https://bazel.build/designs/skyframe.html) and the [Salsa](https://github.com/salsa-rs/salsa) framework: users define logic using a particular API, and the framework manages tracking the dependencies between nodes in a runtime graph.
+
+In order to maximize the amount of work that can be reused at runtime, Pants statically computes the memoization keys for the nodes of the runtime graph from the user specified `@rules` during startup: this process is known as "rule graph construction". See the `Goals` section for more information on the strategy and reasoning for this.
+
+Concepts used in compilers, including live variable analysis and monomorphization, can also be useful in rule graph construction to minimize rule identities and pre-decide which versions of their dependencies they will use. See the "Open Issues" section for a call for assistance with improving the implementation of rule graph construction!
+
+## Concepts
+
+A successfully constructed `RuleGraph` contains a graph where nodes have one of three types, `Rule`s, `Query`s, and `Param`s, which map fairly closely to what a Pants `@rule` author consumes.
+
+### Rules
+
+A `Rule` is a function or coroutine with all of its inputs declared as part of its type signature. The end user type signature is made up of:
+1. the return type of the `Rule`
+2. the positional arguments to the `Rule`
+3. a set of `Get`s which declare the runtime requirements of a coroutine, of the form `Get(output_type, input_type)`
```

Comment:
The reader may not intuit what a `Get` is. It's worth elaborating that a `Rule` can request further input at runtime, by yielding a `Get` object back to the engine, so these also represent part of the input signature, along with the positional arguments, yadda yadda.

Reply:
☝️

Reply:
I think that that is self explanatory by virtue of being part of the signature... not sure how to clarify it.